### PR TITLE
Instruction tuning

### PR DESCRIPTION
**Pull request by AleHD**
Support instruction tuning.

Tasks to do before merging:
- [x] Add preprocess_data.py to tokenize instruction datasets.
- [x] Add instruction tuning loading support in megatron.
- [x] Add proper loss calculation.
- [x] Currently, we only load batches with size `seq_len` but when rolling the tokens to get the labels, we mask the loss of the very last token. We should load sequences of size `seq_len + 1` instead to return batches of size `seq_len` exactly.
- [x] Test correctness of implementation (empirically tested).
- [x] Add a documentation page to explain how to use this.

Thanks to @andreaskoepf and [his fork](https://github.com/andreaskoepf/epfl-megatron/tree/local_changes/) for serving as inspiration to this implementation.

### CODE REVIEW

File: docs/guide/instruction_tuning.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
+# Instruction finetuning
+
+This tutorial will guide you through the basics of instruction finetuning using the Megatron-LLM codebase, using LLaMa 2 7B as the base network.
```
Comment (martinjaggi) [inline docs/guide/instruction_tuning.md:3]: add that these same scripts will also work for larger models such as the 70b llama or falcon

**Review by martinjaggi — APPROVED** [review]

File: finetune.py [hunk 2]
```diff
@@ -48,14 +55,50 @@ def model_provider(pre_process: bool = True, post_process: bool = True):
     return model
 
 
+##
+# Dataset utilities
+##
+
+def get_attention_mask_and_position_ids(data, attention_mask):
+    """Build masks and position id for left to right model."""
```
Comment (martinjaggi) [inline finetune.py:63]: say which kind of attention mask it's meant to do

File: finetune.py [hunk 3]
```diff
@@ -65,21 +108,63 @@ def get_batch(data_iterator):
...
-        tokenizer.eod,
-        args.reset_position_ids,
-        args.reset_attention_mask,
-        args.eod_mask_loss)
+    tokens = data_b["text"]
+    labels = tokens[:, 1:].contiguous()
+    tokens = tokens[:, :-1].contiguous()
+    if args.data_type == "gpt":
+
+        # Get the masks and postition ids.
```
Comment (martinjaggi) [inline finetune.py:116]: typo

**Review by martinjaggi — APPROVED** [review]

File: finetune.py (hunk 3 above)
Comment (AleHD) [reply finetune.py:116]: Mmm I can't see any typo. Maybe it was in another line?

File: docs/guide/instruction_tuning.md (hunk 1 above)
Comment (AleHD) [reply docs/guide/instruction_tuning.md]: Done
